samply/src/shared/symbol_precog.rs

```diff
@@ -0,0 +1,386 @@
+            type Value = PrecogSymbolInfo;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("struct PrecogSymbolInfo")
+            }
+
+            fn visit_map<A: serde::de::MapAccess<'de>>(
+                self,
+                mut map: A,
+            ) -> Result<Self::Value, A::Error> {
+                let mut string_table: Option<StringTable> = None;
+                let mut data: Option<Vec<PrecogLibrarySymbols>> = None;
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        "string_table" => {
+                            string_table = Some(map.next_value()?);
+                        }
+                        "data" => {
+                            data = Some(map.next_value()?);
+                        }
+                        _ => {
+                            map.next_value::<serde::de::IgnoredAny>()?;
+                        }
+                    }
+                }
+
+                // Give the shared string table to each PrecogLibrarySymbols
+                let (string_table, mut data) = (string_table.unwrap(), data.unwrap());
+                let string_table = Arc::new(string_table);
+                for lib in &mut data {
+                    lib.string_table = Some(string_table.clone());
+                }
+                Ok(PrecogSymbolInfo { string_table, data })
+            }
+        }
+
+        deserializer.deserialize_map(PrecogSymbolInfoVisitor)
+    }
+}
+
+impl PrecogLibrarySymbols {
+    fn get_string(&self, index: StringTableIndex) -> &str {
+        self.string_table.as_ref().unwrap().get(index)
+    }
+}
+
+impl wholesym::samply_symbols::SymbolMapTrait for PrecogLibrarySymbols {
+    fn debug_id(&self) -> debugid::DebugId {
+        debugid::DebugId::from_str(&self.debug_id).expect("bad debugid")
+    }
+
+    fn symbol_count(&self) -> usize {
+        // not correct but maybe it's OK
+        self.known_addresses.len()
+    }
+
+    fn iter_symbols(&self) -> Box<dyn Iterator<Item = (u32, std::borrow::Cow<'_, str>)> + '_> {
+        let iter = self.known_addresses.iter().map(move |(rva, info)| {
+            (
+                *rva,
+                std::borrow::Cow::Borrowed(self.get_string(info.symbol)),
+            )
+        });
+
+        Box::new(iter)
+    }
+
+    fn lookup_sync(&self, address: wholesym::LookupAddress) -> Option<wholesym::SyncAddressInfo> {
+        match address {
+            wholesym::LookupAddress::Relative(rva) => {
+                for (known_rva, info) in &self.known_addresses {
+                    if *known_rva == rva {
+                        //eprintln!("lookup_sync: 0x{:x} -> {}", rva, info.symbol.0);
+                        return Some(wholesym::SyncAddressInfo {
+                            symbol: wholesym::SymbolInfo {
+                                address: rva,
+                                size: None,
```

Comment:
I'd prefer it if you stored the correct symbol address. Sometimes you have many different addresses in the same function, and we want the front-end to create a shared symbol for those different addresses. Here you're giving them all different symbol addresses, so the front-end will think they're different symbols with the same name.